test/EndToEnd/Tests/Infrastructure/LocalTestEnvironment.cs

```diff
@@ -0,0 +1,167 @@
+        private async Task<string> StartAppContainerAsync(string appName, int appPortOnHost, int daprPortOnHost)
+        {
+            var imageName = $"{containerRegistry}/{appName}:{containerTag}";
+
+            // Pull the image
+            await dockerClient.Images.CreateImageAsync(
+                new DockerModels.ImagesCreateParameters
+                {
+                    FromImage = imageName
+                },
+                null,
+                new Progress<DockerModels.JSONMessage>((m) => logger.LogInformation($"{m.Status}: {m.Progress}, {m.ProgressMessage}"))
+            );
+
+            var containerParams = new DockerModels.CreateContainerParameters
+            {
+                Image = imageName,
+                Name = appName,
+                ExposedPorts = new Dictionary<string, DockerModels.EmptyStruct>
+                {
+                    { $"{appPortOnHost}", new DockerModels.EmptyStruct() }
+                },
+                HostConfig = new DockerModels.HostConfig
+                {
+                    PortBindings = new Dictionary<string, IList<DockerModels.PortBinding>>
+                    {
+                        { $"{appPortOnHost}", new List<DockerModels.PortBinding> { new DockerModels.PortBinding { HostPort = $"{appPortOnHost}" } } }
+                    }
+                },
+                Env = new List<string>
+                {
+                    $"DAPR_HTTP_HOST=host.docker.internal",
+                    $"DAPR_HTTP_PORT={daprPortOnHost}",
+                    $"ASPNETCORE_URLS=http://*:{appPortOnHost}",
+                }
+            };
+
+            // Create the container
+            var response = await dockerClient.Containers.CreateContainerAsync(containerParams);
+
+            if (response.Warnings != null)
+            {
+                foreach (var warning in response.Warnings)
+                {
+                    logger.LogWarning(warning);
+                }
+            }
+
+            // Start the container
+            var started = await dockerClient.Containers.StartContainerAsync(response.ID, new DockerModels.ContainerStartParameters());
+            if (!started)
+            {
+                throw new SystemException($"Failed to start container {response.ID}");
+            }
+
+            return response.ID;
+        }
+
+        /// <summary>
+        /// Starts the Daprd process.
+        /// </summary>
+        /// <param name="daprHttpPort">The port where the Dapr sidecar should listen.</param>
+        /// <param name="appName">The name of the application.</param>
+        /// <param name="appPort">The port where the application is listening.</param>
+        private int StartDaprProcess(int daprHttpPort, string appName, int? appPort)
+        {
+            var componentsPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".dapr", "components");
```

Comment:
This is nice!

This should work in all different OS, right?

Reply:
Yes, that's the idea.